lib/spack/spack/solver/asp.py

```diff
@@ -1943,9 +1958,11 @@ def add_concrete_versions_from_specs(self, specs, origin):
             if version is None or any(v == version for v in self.possible_versions[s.name]):
                 continue
 
-            self.declared_versions[s.name].append(
-                DeclaredVersion(version=version, idx=0, origin=origin)
-            )
+            if require_checksum and not (isinstance(version, vn.GitVersion) and version.is_commit):
+                continue
```

Comment:
Here we have the opportunity to collect errors and raise early during setup, instead of waiting for clingo to have `errors` in the answer set.